samza-azure/src/main/java/org/apache/samza/system/azureblob/avro/AzureBlobOutputStream.java

```diff
@@ -0,0 +1,314 @@
+    if (!byteArrayOutputStream.isPresent()) {
+      throw new IllegalStateException("Internal Buffer must have been released earlier for blob " + blobAsyncClient.getBlobUrl().toString());
+    }
+
+    int remainingBytes = len;
+    int offset = off;
+    while (remainingBytes > 0) {
+      int bytesToWrite = Math.min(maxBlockFlushThresholdSize - byteArrayOutputStream.get().size(), remainingBytes);
+      byteArrayOutputStream.get().write(b, offset, bytesToWrite);
+      offset += bytesToWrite;
+      remainingBytes -= bytesToWrite;
+      if (byteArrayOutputStream.get().size() >= maxBlockFlushThresholdSize) {
+        uploadBlockAsync();
+      }
+    }
+    metrics.updateWriteByteMetrics(len);
+  }
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public synchronized void flush() {
+    if (byteArrayOutputStream.isPresent()) {
+      uploadBlockAsync();
+    }
+  }
+
+  /**
+   * This api waits for all pending upload (stageBlock task) futures to finish.
+   * It then synchronously commits the list of blocks to persist the actual blob on storage.
+   * @throws IllegalStateException when
+   *       - when closing an already closed stream
+   * @throws RuntimeException when
+   *       - byteArrayOutputStream.close fails or
+   *       - any of the pending uploads fails or
+   *       - blob's commitBlockList fails
+   */
+  @Override
+  public synchronized void close() {
+
+    if (isClosed) {
+      return;
+    }
+
+    LOG.info("{}: Close", blobAsyncClient.getBlobUrl().toString());
+    try {
+      if (byteArrayOutputStream.isPresent()) {
+        byteArrayOutputStream.get().close();
+      }
+      if (blockList.size() == 0) {
+        return;
+      }
+      CompletableFuture<Void> future =
+          CompletableFuture.allOf(pendingUpload.toArray(new CompletableFuture[0]));
+
+      LOG.info("Closing blob: {} PendingUpload:{} ", blobAsyncClient.getBlobUrl().toString(), pendingUpload.size());
+
+      future.get((long) flushTimeoutMs, TimeUnit.MILLISECONDS);
+      LOG.info("For blob: {} committing blockList size:{}", blobAsyncClient.getBlobUrl().toString(), blockList.size());
+      metrics.updateAzureCommitMetrics();
+      Map<String, String> blobMetadata = Collections.singletonMap(BLOB_RAW_SIZE_BYTES_METADATA, Long.toString(totalUploadedBlockSize));
+      blobAsyncClient.commitBlockListWithResponse(blockList, null, blobMetadata, null, null).block();
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      String msg = String.format("Close blob %s failed with exception. Total pending sends %d",
+          blobAsyncClient.getBlobUrl().toString(), pendingUpload.size());
+      throw new RuntimeException(msg, e);
+    } catch (Exception e) {
+      String msg = String.format("Close blob %s failed with exception. Resetting the system producer. %s",
+          blobAsyncClient.getBlobUrl().toString(), e.getLocalizedMessage());
+      throw new RuntimeException(msg, e);
```

Comment:
Specific exception types can be caught within code. It can make logs slightly more readable or searchable to have this more specific type as well.
Do you have a concern with using a more specific exception type?

Reply:
You maybe you could have just one new exception type for any azure blob issue. Any new writers could use that one new exception type. Also note that there is a `SystemProducerException`. Maybe that could be useful in some places (not sure you would use it in this class, but maybe you could use it in your producer class?).